unittests/Dialect/SPIRV/SerializationTest.cpp

```diff
@@ -0,0 +1,237 @@
+        break;
+      }
+      case spirv::Opcode::OpTypeInt: {
+        ASSERT_TRUE(wordCount == 4) << "OpTypeInt must have 3 operands";
+        // Skip signedness.
+        ASSERT_TRUE(binary[currOffset + 2] == 32)
+            << "Test machinery supports only 32 bit width type for OpTypeInt";
+        typeMapID[binary[currOffset + 1]] = opBuilder.getIntegerType(32);
+        break;
+      }
+      case spirv::Opcode::OpTypeStruct: {
+        ASSERT_TRUE(wordCount >= 3) << "Test machinery supports OpTypeStruct "
+                                       "with at least 2 operands";
+        llvm::SmallVector<Type, 0> types;
+        for (auto memberOffset : llvm::seq<uint32_t>(2, wordCount)) {
+          auto memberType = typeMapID.lookup(binary[currOffset + memberOffset]);
+          ASSERT_TRUE(memberType) << "Cannot find member type for OpTypeStruct";
+          types.push_back(memberType);
+        }
+        typeMapID[binary[currOffset + 1]] = spirv::StructType::get(types);
+        break;
+      }
+      default:
+        break;
+      }
+      currOffset += wordCount;
+    }
+  }
+
+protected:
+  MLIRContext context;
+  spirv::ModuleOp module;
+  SmallVector<uint32_t, 0> binary;
+  uint32_t currOffset = 0;
+  llvm::DenseMap<uint32_t, Type> typeMapID;
+  llvm::DenseMap<uint32_t, uint32_t> constantMap;
+  llvm::SmallVector<uint32_t, 16> blockDecorations;
+};
+
+//===----------------------------------------------------------------------===//
+// Block decoration
+//===----------------------------------------------------------------------===//
+
+TEST_F(SerializationTest, BlockDecorationTest) {
+  auto opBuilder = initOpBuilder();
+  // !spv.struct<!spv.array<16 x i32>> type.
+  auto structArray16IntType =
+      createStructArrayType(opBuilder.getIntegerType(32), 16);
+  // !spv.struct<!spv.array<16 x !spv.struct<!spv.array<16 x f32>>>> type.
+  auto structArray16FloatType =
+      createStructArrayType(opBuilder.getF32Type(), 16);
+  auto structNested16StructType =
+      createStructArrayType(structArray16FloatType, 16);
+  // !spv.array<16 x i32> type.
+  auto Array16IntType = spirv::ArrayType::get(opBuilder.getIntegerType(32), 16);
+  // i32 type.
+  auto Int32Type = opBuilder.getIntegerType(32);
+  // !spv.struct<!spv.array<32 x i32>> type.
+  auto structArray32IntType =
+      createStructArrayType(opBuilder.getIntegerType(32), 32);
+
+  // Add globals.
+  addGlobalVar(opBuilder, "var0", structArray16IntType,
+               spirv::StorageClass::Uniform);
+  addGlobalVar(opBuilder, "var1", structNested16StructType,
+               spirv::StorageClass::StorageBuffer);
+  addGlobalVar(opBuilder, "var2", Array16IntType, spirv::StorageClass::Uniform);
+  addGlobalVar(opBuilder, "var3", Int32Type,
+               spirv::StorageClass::StorageBuffer);
+  addGlobalVar(opBuilder, "var4", structArray32IntType,
+               spirv::StorageClass::Input);
+
+  // Serialize module.
+  ASSERT_FALSE(failed(spirv::serialize(module, binary)));
```

Comment:
I'm not a huge fan of the way this test is structured; what is this actually testing? It seems like in these serializer tests you have to re-create parts of the deserializer, which requires a bunch of additional logic, verification and potential bugs that seem unrelated to the feature being tested. Is there a way to restructure this, hopefully without using c++ unit tests that still achieves the same goal?

Reply:
@River707 thanks for review, the purpose of this test to is check the some parts of the binary form which does not affect IR, I agree that looks bad; the discussion of the unittest of  binary form was here https://github.com/tensorflow/mlir/issues/77 , but we didn't discuss the realization, it might be better to collect Block decorated types in "deserilizer", but I'm not sure about it.
Here is a simple example of Block decoration:
`OpDecorate %_struct_6 Block`
```

; SPIR-V
; Version: 1.3
; Generator: Khronos SPIR-V Tools Assembler; 0
; Bound: 20
; Schema: 0
               OpMemoryModel Logical GLSL450
               OpName %var0 "var0"
               OpMemberDecorate %_struct_6 0 Offset 0
               OpDecorate %_struct_6 Block
        %int = OpTypeInt 32 1
     %int_16 = OpConstant %int 16
%_arr_int_int_16 = OpTypeArray %int %int_16
  %_struct_6 = OpTypeStruct %_arr_int_int_16
       %var0 = OpVariable %_ptr_Uniform__struct_6 Uniform
```


`%_struct_6` is a `<id> `and we have to deduce its type step by step:

` %_struct_6 = OpTypeStruct %_arr_int_int_16` 
`%_arr_int_int_16 = OpTypeArray %int %int_16` 
` %int_16 = OpConstant %int 16`
`%int = OpTypeInt 32 1`

Reply:
@antiagainst thanks for detailed explanation, fixed regarding to you comment.